File: src/main/java/com/felipe/booking/domain/usecase/ValidateBookingUseCase.java
package com.felipe.booking.domain.usecase;

import com.felipe.booking.domain.model.Booking;
import com.felipe.booking.domain.validations.BookingValidator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Set;

@Component
public class ValidateBookingUseCase {

    @Autowired
    private Set<BookingValidator> validators;

    public Mono<Booking> validate(Booking booking) {
        return Mono.just(booking)
                .flatMap(book -> executeValidations(booking));
    }

    private Mono<Booking> executeValidations(Booking booking) {
     return Flux.fromIterable(validators)
                .flatMap ( validator -> validator.validate(booking) )
                .collectList();
    }
}

File: src/main/kotlin/com/felipe/booking/alten/domain/usecase/ValidateBookingUseCase.kt
package com.felipe.booking.alten.domain.usecase

import com.felipe.booking.alten.domain.model.Booking
import com.felipe.booking.alten.domain.validations.BookingValidator
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Component
class ValidateBookingUseCase(
    private val validators: Set<BookingValidator<Booking>>
) {

    fun validate(booking: Booking) =
        Mono.just(booking)
            .flatMap { executeValidations(booking) }

    private fun executeValidations(booking: Booking) =
        Flux.fromIterable(validators)
            .flatMap { it.validate(booking) }
            .collectList()
}